mod/cp_notifications/lib/functions.php

```diff
@@ -373,30 +442,51 @@ function create_digest($invoked_by, $subtype, $entity, $send_to, $entity_url = '
 				'content_author_url' => $invoked_by->getURL()
 			);
 
-			if (!is_array($digest_collection['personal']['cp_mention']))
-				$digest_collection['personal']['cp_mention'] = array();
-
-			$digest_collection['personal']['cp_mention'][] = json_encode($content_array);
-
+			$entity_guid = $entity->guid;
+			$user_guid = $send_to->getGUID();
+			$entry_type = 'personal';
+			$group_name = NULL;
+			$action_type = 'mention';
+			$notification_entry = json_encode($content_array);
 			break;
 
 		default:
 
 			$entity = get_entity($entity->guid);
 			
-			if ($entity->getContainerEntity() instanceof ElggGroup)
-				$digest_collection['group']["<a href='{$entity->getContainerEntity()->getURL()}?utm_source=notification_digest&utm_medium=email'>{$entity->getContainerEntity()->name}</a>"]['new_post'][$entity->guid] = json_encode($content_array);
-			else 
-				$digest_collection['personal']['new_post'][$entity->guid] = json_encode($content_array);
+			if ($entity->getContainerEntity() instanceof ElggGroup) {
+			
+				$entity_guid = $entity->guid;
+				$user_guid = $send_to->getGUID();
+				$entry_type = 'group';
+				$group_name = "<a href='{$entity->getContainerEntity()->getURL()}?utm_source=notification_digest&utm_medium=email'>{$entity->getContainerEntity()->name}</a>";
+				$action_type = 'new_post';
+				$notification_entry = json_encode($content_array);
+
+			}
+			else {
+				$entity_guid = $entity->guid;
+				$user_guid = $send_to->getGUID();
+				$entry_type = 'personal';
+				$group_name = NULL;
+				$action_type = 'new_post';
+				$notification_entry = json_encode($content_array);
+			}
 
 			break;
 	}
 
-	// save the information to the digest object for later access
-	$ia = elgg_set_ignore_access(true);
-	$digest->description = json_encode($digest_collection);
-	$digest->save();
-	elgg_set_ignore_access($ia);
+	$group_name = base64_encode($group_name);
+
+	/// check if record exists already, if not then proceed.
+	$query = "SELECT 1 FROM notification_digest WHERE entity_guid = {$entity_guid} AND user_guid = {$user_guid} AND notification_entry = '{$notification_entry}'  LIMIT 1";
```

Comment:
pretty sure user_guid should be first, so maybe (user_guid, notification_entry, entity_guid) ? I think that should cover everything actually.